### PR TITLE
Unit test for save history feature 1818

### PR DESCRIPTION
**Pull request by lzxm160**
work on #1818

### CODE REVIEW

**Comment by dustinxie** [comment]
i think we need to first adapt the AccountState() func, allow it to access state at diff height, by pass in height as an extra param, or using option (so can avoid changing existing code)

add test code for statetx.go (for account) and trie.go (for contract), that is unit test

then add test in blockchain.go (which is what this PR did), that is considered integration or e2e test

also suggest we can leverage existing test code instead of adding new code like this PR. for example, TestLoadBlockchainfromDB() calls addTestingTsfBlocks(), which already has code to commit couple of blocks to a local blockchain. You can maybe add another test into TestLoadBlockchainfromDB() with saveHistory option to test reading diff states at diff height

**Comment by CoderZhi** [comment]
let's don't test against statetx, but workingset. statetx is not designed to store archive data.

**Comment by dustinxie** [comment]
> let's don't test against statetx, but workingset. statetx is not designed to store archive data.

statetx needs to be tested as well, it stores account's state at diff block heights

@lzxm160  we'll need to add into statetx.go new API to read account's state at certain block height

**Comment by lzxm160** [comment]
> o read account's state at certain block heigh

already change to workset,not using statetx

**Comment by lzxm160** [comment]
> > let's don't test against statetx, but workingset. statetx is not designed to store archive data.
> 
> statetx needs to be tested as well, it stores account's state at diff block heights
> 
> @lzxm160 we'll need to add into statetx.go new API to read account's state at certain block height

already change to workset,not using statetx

**Comment by lzxm160** [comment]
need to fix mock later,cannot install mockgen in local vm because of the gfw

**Comment by lzxm160** [comment]
need to merge with latest master and test

File: state/factory/statetx.go [hunk 1]
```diff
@@ -232,6 +232,23 @@ func (stx *stateTX) State(hash hash.Hash160, s interface{}) error {
...
+	}
+	data, err := ri.Get(height)
+	if err != nil {
+		if errors.Cause(err) == db.ErrNotExist {
+			return errors.Wrapf(state.ErrStateNotExist, "state of %x doesn't exist", addr)
+		}
+		return errors.Wrapf(err, "error when getting the state of %x", addr)
+	}
+	return state.Deserialize(s, data)
+}
```
Comment (CoderZhi) [inline state/factory/statetx.go]: as commented previously, we don't need to implement this one. Instead, we should throw an error "not supported". StateTX and StateDB are not designed for archive mode.

File: state/factory/factory_test.go [hunk 2]
```diff
@@ -303,6 +303,22 @@ func TestState(t *testing.T) {
 	testState(sf, t)
 }
 
+func TestHistoryState(t *testing.T) {
+	testTrieFile, _ := ioutil.TempFile(os.TempDir(), triePath)
+	cfg := config.Default
+	cfg.Chain.TrieDBPath = testTrieFile.Name()
+	cfg.Chain.EnableHistoryStateDB = true
```
Comment (CoderZhi) [inline state/factory/factory_test.go]: add a unit test: if this is false, we should expect an error, say ErrNoArchiveData

**Pull request merged**